app/uk/gov/hmrc/gform/sharedmodel/formtemplate/Expr.scala

```diff
@@ -103,6 +103,7 @@ object FormCtx {
 final case class ParamCtx(queryParam: QueryParam) extends Expr
 final case class AuthCtx(value: AuthInfo) extends Expr
 final case class UserCtx(value: UserField) extends Expr
+final case class UserFuncCtx(value: UserField, func: UserFieldFunc) extends Expr
```

Comment:
This assume that `func` can be applied to any product from `value` type. Is that correct? For example do we want to ask `affinityGroup.count` or `enrolledIdentifier.count`?

Reply:
It does say so looking at the type definition. I dont about AffinityGroup or EnrolledIdentifier. Alternatively, to be precise, we could rename the class to

```final case class UserEnrollmentFuncCtx(enrolment: UserField.Enrolment, func: UserFieldFunc) extends Expr```

Reply:
Actually, when using the above, derived is unable to build a OFormat[_] for UserField class. This has happened in the past for other Expr types too

```
/Users/girishkamath/development/hmrc/gform/app/uk/gov/hmrc/gform/sharedmodel/formtemplate/Expr.scala:117:55
could not find Lazy implicit value of type julienrf.json.derived.DerivedReads[A]
Error occurred in an application involving default arguments.
  implicit val format: OFormat[Expr] = derived.oformat()
```

Reply:
Problems aside I'm not in favour of this definition (ie. defining new `Expr` for something which is not new) 
```
final case class UserEnrollmentFuncCtx(enrolment: UserField.Enrolment, func: UserFieldFunc)
```
I'd rather express it like this
```
final case class Enrolment(serviceName: ServiceName, identifierName: IdentifierName, func: UserFieldFunc)
      extends UserField
```
We can introduce synthetic function `Print` to express situation when no function is defined.

Reply:
Why not introduce a new UserField and avoid Print alltogether. This synthetic function is being forced upon with the approach

```final case class EnrolmentFunc(serviceName: ServiceName, identifierName: IdentifierName, func: UserFieldFunc)
      extends UserField```

or 

```final case class EnrolmentFunc(enrolment: UserField.Enrolment, func: UserFieldFunc)
      extends UserField```

Reply:
Introducing new `UserField` is introducing more friction. 

And in such a case whole `final case class Enrolment` would be synthetic.

If you don't like synthetic function representation `case object Print extends UserFieldFunc` we can be less expressive and make `UserFieldFunc` optional in `Enrolment` instead:
```
final case class Enrolment(serviceName: ServiceName, identifierName: IdentifierName, func: Option[UserFieldFunc])
      extends UserField
```

Reply:
Thanks, I have made it optional. Please re-review